components/merkledb/src/migration.rs

```diff
@@ -450,10 +524,53 @@ impl MigrationHelper {
     /// Use [`flush_migration`] to flush the migrated data.
     ///
     /// [`flush_migration`]: fn.flush_migration.html
-    pub fn finish(self) -> crate::Result<Hash> {
-        let patch = self.fork.into_patch();
+    pub fn finish(self) -> Result<Hash, MigrationError> {
+        let patch = self.fork.unwrap().into_patch();
         let hash = Migration::new(&self.namespace, &patch).state_hash();
-        self.db.merge(patch).map(|()| hash)
+        if self.abort_handle.load(Ordering::SeqCst) {
+            Err(MigrationError::Aborted)
+        } else {
+            self.db.merge(patch).map_err(MigrationError::Merge)?;
+            Ok(hash)
+        }
+    }
+}
+
+/// Errors emitted by `MigrationHelper` methods.
+#[derive(Debug, Fail)]
+pub enum MigrationError {
+    /// Failed to merge migration changes to database.
+    #[fail(display = "Failed to merge migration changes to database: {}", _0)]
+    Merge(#[fail(cause)] crate::Error),
+
+    /// Migration has been aborted.
+    #[fail(display = "Migration was aborted")]
+    Aborted,
+}
+
+/// Handle allowing to signal to `MigrationHelper` that the migration has been aborted.
+/// Signalling is performed on handle drop, unless it is performed with [`forget`](#method.forget)
+/// method.
+#[derive(Debug)]
+pub struct AbortHandle {
+    inner: Arc<AtomicBool>,
+}
+
+impl Drop for AbortHandle {
+    fn drop(&mut self) {
+        self.inner.store(true, Ordering::SeqCst);
+    }
+}
+
+impl AbortHandle {
+    /// Returns `true` if the `MigrationHelper` associated with this handle has been dropped.
+    pub fn is_finished(&self) -> bool {
+        Arc::strong_count(&self.inner) <= 1
+    }
+
+    /// Drops the handle without aborting the migration.
+    pub fn forget(mut self) {
+        self.inner = Arc::new(AtomicBool::default());
```

Comment:
Hm. Can't the implementation be just a no-op? If we're consuming `self`, the `drop` on `Arc` will be called and the result should be the same, or am I missing something?

Reply:
Due to `self.inner` swap, the destructor will be called on an `Arc` that is not connected to a `MigrationHelper`.